src/LibraryManager/Utilities/ParallelUtility.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Web.LibraryManager.Utilities
+{
+    internal static class ParallelUtility
+    {
+        /// <summary>
+        /// Executes an ation over each element of a collection, up to the specified degree of parallelism.
+        /// </summary>
+        /// <param name="action">The action to be applied to each element</param>
+        /// <param name="degreeOfParallelism">How many tasks to run in parallel</param>
+        /// <param name="items">The items to </param>
+        public static Task ForEachAsync<T>(Func<T, Task> action, int degreeOfParallelism, IEnumerable<T> items)
+        {
+            return ForEachAsync<T>(action, degreeOfParallelism, items, CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Executes an ation over each element of a collection, up to the specified degree of parallelism.
+        /// </summary>
+        /// <param name="action">The action to be applied to each element</param>
+        /// <param name="degreeOfParallelism">How many tasks to run in parallel</param>
+        /// <param name="items">The items to </param>
+        /// <param name="cancellationToken">Cancellation token</param>
+        public static async Task ForEachAsync<T>(Func<T, Task> action, int degreeOfParallelism, IEnumerable<T> items, CancellationToken cancellationToken)
+        {
+            action = action ?? throw new ArgumentNullException(nameof(action));
+            items = items ?? throw new ArgumentNullException(nameof(items));
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var semaphore = new SemaphoreSlim(degreeOfParallelism, degreeOfParallelism);
+            var allTasks = new List<Task>();
+
+            foreach (T item in items)
+            {
+                await semaphore.WaitAsync();
+                cancellationToken.ThrowIfCancellationRequested();
+
+                Task task = DoAction(action, item, semaphore);
+                allTasks.Add(task);
+            }
+
+            await Task.WhenAll(allTasks);
+        }
+
+        private static async Task DoAction<T>(Func<T, Task> act, T input, SemaphoreSlim s)
+        {
+            try
+            {
+                await act(input);
```

Comment:
Should this be Task.Run or are we relying on the Func to actually start a new thread? Our current usage is probably fine because async download file code will end up creating a new thread under the covers, but it seems that in general case this would be running on the same thread if your action Func didn't create a new thread, no? As in

https://stackoverflow.com/questions/27265818/does-the-use-of-async-await-create-a-new-thread

Reply:
Task.Run is blocking, so I don't want it here.  It would lead to starving up to `degreeOfParallelism` threads from the thread pool.

I don't care about creating new threads, the thread pool will manage that as the number of queued work items (Tasks) grows.  Since the long pull on the current usage of this is network I/O, being able to reuse threads while they are in an awaiting state is optimal anyways.

Reply:
I mean Run/RunAsync, basically the question is - do we want to make sure we actually run things in parallel, because the current method doesn't actually insure parallelism AFAIK. Our current usage with network I/O will result in parallelism, but in general case it's not.

Reply:
Calling `DoAction(..)` above is already the equivalent of `Task.RunAsync(() => DoAction(...));`  So by this await, we're already running in a separate task from the code path that launched it (probably executing in parallel).

The foreach loop basically queues up as many Tasks as it can (let's go with 10), from a single thread, but it doesn't yield while it's doing so until it gets blocked on the `semaphore.WaitAsync()`.  If the other threads are sitting idle, they'll steal the work from this thread's job queue (that's what the default TaskScheduler does, and I only call out "default" here because it's always possible to write your own and pass it in as a parameter to [TaskFactory.StartNew](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskfactory.startnew?view=netframework-4.8)), and that's how we get them to run in parallel.  There isn't anything we can do to force parallel execution ourselves through async/await.  These language features only provide asynchronicity, not parallelism.  It's entirely up to the TaskScheduler to provide parallel execution (and even using [TaskCreationOptions.LongRunning](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskcreationoptions?view=netframework-4.8) only provides a hint to the scheduler to create a new thread, but doesn't guarantee it).